pymc_marketing/mmm/hsgp.py

```diff
@@ -0,0 +1,967 @@
+#   Copyright 2024 The PyMC Labs Developers
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""HSGP components."""
+
+from __future__ import annotations
+
+from collections.abc import Iterable
+from enum import Enum
+from typing import cast
+
+import numpy as np
+import numpy.typing as npt
+import pymc as pm
+import pytensor.tensor as pt
+import xarray as xr
+from matplotlib.axes import Axes
+from matplotlib.figure import Figure
+from pydantic import BaseModel, Field, InstanceOf, model_validator, validate_call
+from pymc.distributions.shape_utils import Dims
+from pytensor.tensor import TensorLike
+from pytensor.tensor.variable import TensorVariable
+from typing_extensions import Self
+
+from pymc_marketing.plot import SelToString, plot_curve
+from pymc_marketing.prior import Prior, create_dim_handler
+
+
+@validate_call
+def create_complexity_penalizing_prior(
+    *,
+    alpha: float = Field(0.1, gt=0, lt=1),
+    lower: float = Field(1.0, gt=0),
+) -> Prior:
+    R"""Create prior that penalizes complexity for GP lengthscale.
+
+    The prior is defined with the following property:
+
+    .. math::
+
+        P[\ell < \text{lower}] = \alpha
+
+    Where :math:`\ell` is the lengthscale
+
+    Parameters
+    ----------
+    alpha : float
+        Tail probability.
+    lower : float
+        Lower bound for the lengthscale.
+
+    Returns
+    -------
+    Prior
+        Weibull prior for the lengthscale with the given properties.
+
+    References
+    ----------
+    .. [1] Geir-Arne Fuglstad, Daniel Simpson, Finn Lindgren, Håvard Rue (2015).
+
+    """
+    lam_ell = -pt.log(alpha) * (1.0 / pt.sqrt(lower))
+
+    return Prior(
+        "Weibull",
+        alpha=0.5,
+        beta=1.0 / pt.square(lam_ell),
+        transform="reciprocal",
+    )
+
+
+def create_constrained_inverse_gamma_prior(
+    *,
+    upper: float,
+    lower: float = 1.0,
+    mass: float = 0.9,
+) -> Prior:
+    """Create a lengthscale prior for the HSGP model.
+
+    Parameters
+    ----------
+    upper : float
+        Upper bound for the lengthscale.
+    lower : float
+        Lower bound for the lengthscale. Default is 1.0.
+    mass : float
+        Mass of the lengthscales. Default is 0.9 or 90%.
+
+    Returns
+    -------
+    Prior
+        The prior for the lengthscale.
+
+    """
+    return Prior(
+        "InverseGamma",
+    ).constrain(lower=lower, upper=upper, mass=mass)
+
+
+def approx_hsgp_hyperparams(
```

Comment:
Is it the same as in https://github.com/pymc-devs/pymc/blob/main/pymc/gp/hsgp_approx.py#L97 ?

Reply:
Not in the min pymc version that we support. How should we handle?